python/cuml/linear_model/logistic_regression_mg.pyx

```diff
@@ -0,0 +1,198 @@
+#
+# Copyright (c) 2019-2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# distutils: language = c++
+
+from cuml.internals.safe_imports import gpu_only_import
+cp = gpu_only_import('cupy')
+from cuml.internals.safe_imports import cpu_only_import
+np = cpu_only_import('numpy')
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+
+from cuml.common import input_to_cuml_array
+import numpy as np
+
+from cuml.internals.array import CumlArray
+from cuml.internals.api_decorators import device_interop_preparation
+from cuml.linear_model import LogisticRegression 
+from cuml.solvers.qn import QN
+from cuml.solvers.qn import QNParams
+
+from pylibraft.common.handle cimport handle_t
+
+from cuml.solvers.qn import __is_col_major
+
+# the cdef was copied from cuml.linear_model.qn
+cdef extern from "cuml/linear_model/glm.hpp" namespace "ML::GLM" nogil:
+
+    cdef enum qn_loss_type "ML::GLM::qn_loss_type":
+        QN_LOSS_LOGISTIC "ML::GLM::QN_LOSS_LOGISTIC"
+        QN_LOSS_SQUARED  "ML::GLM::QN_LOSS_SQUARED"
+        QN_LOSS_SOFTMAX  "ML::GLM::QN_LOSS_SOFTMAX"
+        QN_LOSS_SVC_L1   "ML::GLM::QN_LOSS_SVC_L1"
+        QN_LOSS_SVC_L2   "ML::GLM::QN_LOSS_SVC_L2"
+        QN_LOSS_SVR_L1   "ML::GLM::QN_LOSS_SVR_L1"
+        QN_LOSS_SVR_L2   "ML::GLM::QN_LOSS_SVR_L2"
+        QN_LOSS_ABS      "ML::GLM::QN_LOSS_ABS"
+        QN_LOSS_UNKNOWN  "ML::GLM::QN_LOSS_UNKNOWN"
+
+    cdef struct qn_params:
+        qn_loss_type loss
+        double penalty_l1
+        double penalty_l2
+        double grad_tol
+        double change_tol
+        int max_iter
+        int linesearch_max_iter
+        int lbfgs_memory
+        int verbose
+        bool fit_intercept
+        bool penalty_normalized
+
+
+cdef extern from "cuml/linear_model/qn_mg.hpp" namespace "ML::GLM::opg" nogil:
+
+    void qnFit(
+        const handle_t& handle,
+        const qn_params& pams, 
+        float *X,
+        bool X_col_major,
+        float *y,
+        int N,
+        int D,
+        int C,
+        float *w0,
+        float *f,
+        int *num_iters,
+        int n_samples,
+        int rank,
+        int n_ranks) except +
+
+
+class LogisticRegressionMG(LogisticRegression):
```

Comment:
My suggestion is to use the design from `LinearRegressionMG` as much as possible to model the relationship between `LogisticRegression` and `LogisticRegressionMG`. There are some more classes and mixins that can reduce boilerplate further.

Reply:
Find it a bit challenging to inherit mixins, for two reasons:

(1) min_lbfgs accepts float*.  mixins _fit accepts vector<float*>
(2) min_lbfgs supports multiple classes so coef_ can be a matrix.  mixins assumes coef_ to be a vector of D + pams.fit_intercept length. 

Any ideas to simplify the implementation?

Reply:
(3) Another issue pops up that relates to mixins._fit and (1). It seems mixins.fit converts input vectors and labels into a special type of points (i.e. X_arg, y_arg) using opg.build_data_t(X_arys). Is there s way to convert X_arg and y_arg to float* (min_lbfgs requires the float*)?

Reply:
Combined the design from 'LinearRegressionMG' and singe-gpu 'LogisticRegression'. Specifically:
(1) Got self._coef initialization support LinearRegressionMG style (i.e. 1-D array) and LogisticRegression style (i.e. 2-D array). 
(2) Converted type of input vectors and input labels) from LienarRegressionMG style (vector of Matrix:Data<float> type) to LogisticRegression style (float*). 
(3) set n_classes to 2 in this PR and will work on multiple class in following PRs.